autoload/ale/c.vim

```diff
@@ -22,6 +23,84 @@ function! ale#c#FindProjectRoot(buffer) abort
     return ''
 endfunction
 
+function! ale#c#ParseCFlagsToList(buffer, cflags) abort
+    let l:project_root = ale#c#FindProjectRoot(a:buffer)
+    let l:previous_option = ''
+    let l:shell_option = 0
+    let l:macro_option = 0
+    let l:cflags_list = []
+
+    for l:option in a:cflags
+        " Check if cflag contained spaces
+        if l:shell_option || stridx(l:option, '=`') >= 0
+            " Cflag contained shell command with spaces (ex. -D='date +%s')
+            let l:shell_option = 1
+            let l:previous_option .= l:option
+            if l:option[-1: -1] isnot? '`'
+                let l:previous_option .= ' '
+                continue
+            endif
+            let l:shell_option = 0
+        elseif l:macro_option || stridx(l:option, '$((') > 0
+            " Cflag contained macro with spaces (ex -Da=$(( 4 * 20  )))
+            let l:macro_option = 1
+            let l:previous_option .= l:option
+            if stridx(l:option, '))') < 0
+                let l:previous_option .= ' '
+                continue
+            endif
+            let l:macro_option = 0
+        endif
+        if l:previous_option isnot? ''
+            let l:option = l:previous_option
+            let l:previous_option = ''
+        endif
+        " Fix relative paths if needed
+        if stridx(l:option, '-I') >= 0
+            if stridx(l:option, '-I' . s:sep) < 0
+                let l:option = '-I' . l:project_root . s:sep . l:option[2:]
+            endif
+        endif
+        " Parse the cflag
+        if stridx(l:option, '-I') >= 0 ||
+           \ stridx(l:option, '-D') >= 0
+            if index(l:cflags_list, l:option) < 0
+                call add(l:cflags_list, l:option)
+            endif
+        endif
+    endfor
+    return l:cflags_list
+endfunction
+
+function! ale#c#ParseCFlags(buffer, stdout_make) abort
+    if g:ale_c_parse_makefile
+        for l:cflags in split(a:stdout_make, '\n')
+            if stridx(l:cflags, expand('#' . a:buffer . '...'))
+                let l:cflags = split(l:cflags)
+                break
+            endif
+        endfor
+        if !empty(l:cflags)
+            return ale#c#ParseCFlagsToList(a:buffer, l:cflags)
+       endif
+    endif
+    retur []
+endfunction
+
+function! ale#c#ParseMakefile(buffer) abort
+    if g:ale_c_parse_makefile
+        let l:project_root = ale#c#FindProjectRoot(a:buffer)
+        let l:project_cflags = []
+
+        if !empty(l:project_root)
+            if !empty(globpath(l:project_root, 'Makefile', 0))
```

Comment:
Instead of using `FindProjectRoot` and then `globpath` to find the file, use `ale#path#FindNearestFile` to find `Makefile` instead. `globpath` can run kind of slowly for large directories, and you should find the nearest `Makefile`, instead of the one at the project root. I believe the Linux kernel and a few other projects used nested Makefiles so you can make subsets of the project faster.